packages/compass-sidebar/src/modules/databases.ts

```diff
@@ -105,64 +118,78 @@ export default function reducer(
 
     return {
       ...state,
-      filterRegex: action.filterRegex,
-      filteredDatabases: filterDatabases(state.databases, action.filterRegex),
-      expandedDbList,
+      [action.connectionId]: {
+        ...state[action.connectionId],
+        filterRegex: action.filterRegex,
+        filteredDatabases: filterDatabases(
+          state[action.connectionId]?.databases ?? [],
+          action.filterRegex
+        ),
+        expandedDbList,
+      },
     };
-  }
-
-  if (action.type === CHANGE_DATABASES) {
+  } else if (action.type === CHANGE_DATABASES) {
     return {
       ...state,
-      databases: action.databases,
-      filteredDatabases: filterDatabases(action.databases, state.filterRegex),
+      [action.connectionId]: {
+        ...state[action.connectionId],
+        databases: action.databases,
+        filteredDatabases: filterDatabases(
+          action.databases,
+          state[action.connectionId]?.filterRegex
+        ),
+      },
     };
   }
+
   return state;
 }
 
-/**
- * The change databases action creator.
- *
- * @param {Array} databases
- *
- * @returns {Object} The action.
- */
-export const changeDatabases = (databases: Database[]) => ({
+export const changeDatabases = (
+  connectionId: ConnectionInfo['id'],
+  databases: Database[]
+) => ({
   type: CHANGE_DATABASES,
+  connectionId,
   databases,
 });
 
 export const toggleDatabaseExpanded =
   (
-    id: string,
+    connectionId: ConnectionInfo['id'],
+    databaseId: string,
     forceExpand: boolean
   ): SidebarThunkAction<void, DatabasesAction> =>
   (dispatch, getState, { globalAppRegistry }) => {
-    const { database } = toNS(id);
-    const { databases } = getState();
-    const expanded = forceExpand ?? !databases.expandedDbList[database];
+    const { database } = toNS(databaseId);
+    const { expandedDbList } = getState().databases[connectionId];
+    const expanded = forceExpand ?? !expandedDbList[database];
```

Comment:
You may want to be careful here, afaict `getState().databases` and `expandedDbList` are not initialized as null-prototype objects anymore, so e.g. a database called `toString` would always be expanded now (something the previous code handled)?

Reply:
So you mean that because the initial state is not null, if a database is called toString it will be always expanded? Just to understand.

To be fair I didn't see any code handling this specific issue before, but I might missed it.

Reply:
> So you mean that because the initial state is not null, if a database is called toString it will be always expanded? 

Err, sorry, I mixed that up because of the toggling – it's _never_ expanded here unless `forceExpand` is set, right? Because `!expandedDbList['toString']` will evaluate to `false` because `expandedDbList` now inherits from `Object` (unless I'm missing something)

> To be fair I didn't see any code handling this specific issue before, but I might missed it.

This is what I was referring to: https://github.com/mongodb-js/compass/blob/4ae8d14e4e2d5333a7f9ace82d7d01910b2f0a3a/packages/compass-sidebar/src/modules/databases.ts#L61

Reply:
Hmm, I get it now, but wouldn't it recreate a normal object when we do Object.fromEntries?

https://github.com/mongodb-js/compass/blob/4ae8d14e4e2d5333a7f9ace82d7d01910b2f0a3a/packages/compass-sidebar/src/modules/databases.ts#L100

And also when shallow copying here:

https://github.com/mongodb-js/compass/blob/4ae8d14e4e2d5333a7f9ace82d7d01910b2f0a3a/packages/compass-sidebar/src/modules/databases.ts#L81

Asking because I don't know.

Reply:
Yeah, it would ... sigh. :) The shallow copy at least should have probably included a `__proto__: null`.

Anyway, yeah, nothing this PR breaks then. Let's merge it, I can do a small follow-up to fix this :)

Reply:
I have a task for this kind of small follow ups. No worries, I'll add this topic there so we can fix it later haha.